### PR TITLE
Traffic-Based App Rewards: Enable app reward calculation on TestNet bootstrap without featured-apps

### LINKED ISSUE #5624: Support bootstrapping network with traffic-based rewards enabled
**Note:** This one focuses only on adding a single integration test that shows that a network can be bootstrapped with TBAR enabled using a four SV topology.

We need add support for computing round 0 activity totals
Ref [this discussion](https://github.com/canton-network/splice/pull/5552#discussion_r3265723328) about the need for this support for TestNet reset and use of traffic-based rewards as default in integration tests.

### PR DESCRIPTION
**Pull request by adetokunbo**
Fixes #5624

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If an upgrade test is required, comment `/upgrade_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a logical synchronizer upgrade test is required (from canton-3.5), comment `/lsu_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/canton-network/splice/blob/main/docs/src/release_notes.rst).
- [x] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/canton-network/splice/blob/main/TESTING.md#running-and-debugging-integration-tests) or use your favorite screenshot tool

#### Merge Guidelines
- [x] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

### CODE REVIEW

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala [hunk 1]
```diff
@@ -284,26 +279,25 @@ class DbAppActivityRecordStore(
...
-      } else None
+    val earliestRound = if (items.nonEmpty) {
+      items
+        .map(_.roundNumber)
+        .foldLeft(Long.MaxValue)(math.min)
+    } else {
+      // No activity records (e.g., no featured app providers).
+      // Use lastArchivedRound as earliest so rounds after it are
+      // considered to have complete (empty) activity.
+      lastArchivedRoundO.getOrElse(-1L)
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala]: The `lastArchivedRoundO` is a value obtained from the scan's own store, so in abscence of the data we will always put `-1`, is this ok?

Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala]: It is also not clear if the insertion is guarded against the absence of traffic-summaries causing no `items`

Comment (adetokunbo) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala]: > is this ok? 

No.  I took out the test on isFirstSV prematurely.  I've put it back and updated the logic here to make it clearer:
- Has activity records → use their min round as earliestRound, create meta row                                                                                                                                                             
- No records but has lastArchivedRoundO → use it as earliestRound, create meta row (handles the "no featured apps but rounds are advancing" case)
- No records, no lastArchivedRoundO, isFirstSv → use fallback values (earliestRound = -1, lastArchived = 0), create meta row (bootstraps round 0)                                                                                          
- No records, no lastArchivedRoundO, not isFirstSv → skip meta creation, a later batch will have the info

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala [hunk 2]
```diff
@@ -284,49 +280,66 @@ class DbAppActivityRecordStore(
...
+
+    // earliestRound: the lowest round covered by this ingestion batch.
+    //   - From activity records when present
+    //   - From lastArchivedRound when no featured apps produced records
+    //   - From bootstrap (-1) on a fresh firstSV with no archived rounds
+    val earliestRound = items
+      .map(_.roundNumber)
+      .minOption
+      .orElse(lastArchivedRoundO)
+      .orElse(if (isFirstSv) Some(-1L) else None)
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:331]: Here if on an existing network, on the `isFirstSv`, the `lastArchivedRoundO` is none and there are no items, the `-1` would still be inserted. I expect the earliestRound = -1 value to be based on some evidence that we are actually on round 0.

Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:331]: This may not be worse than the existing logic, but lets consider what is blocking us to have an evidence for round 0 while doing the insert here.

Comment (adetokunbo) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:331]: The bootstrap path (`earliestRound = -1`, `lastArchived = 0`) only fires when all three conditions hold simultaneously:

1. `isFirstSv` is true
2. No prior meta rows exist in the DB (`maxVersions.isEmpty` in `ensureMetaDBIO`)
3. `lastArchivedRoundO` is `None`

On an existing network after restart, condition 2 fails — the meta row persists in the DB across restarts. On an existing network where rounds are advancing, condition 3 fails — `lookupLatestArchivedOpenMiningRound` returns `Some(...)`.

The only scenario where all three hold is a truly fresh founding SV that has never ingested activity before and has no archived rounds — i.e., bootstrap. Once real data arrives, `updateLastArchivedRoundDBIO` overwrites the bootstrap values.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala [hunk 3]
```diff
@@ -284,49 +280,66 @@ class DbAppActivityRecordStore(
...
+
+    // lastArchived: the highest round archived as of this verdict batch.
+    //   - From the caller when available
+    //   - Bootstrapped to 0 on a fresh firstSV to make round 0 complete
+    val lastArchived = lastArchivedRoundO
```
Comment (dfordivam) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:338]: I expect that `lastArchivedRoundO` should be `Some 0` and this isn't needed, did we see a scenario where this was necessary

Comment (adetokunbo) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:338]: On a fresh network, round 0 is open but not yet archived when the first verdicts arrive. `lastArchivedRoundO` comes from `lookupLatestArchivedOpenMiningRound(maxRecordTime)`, which queries the archive table for `OpenMiningRound` contracts — at bootstrap there are none, so it returns `None`.

This is the circular dependency: round 0 can't close without reward totals, reward totals need the meta row with a non-NULL `last_archived_round`, and `last_archived_round` needs a round to have been archived. The fallback breaks this cycle.

Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:338]: hmm.  there is no circular dependency, ie round 0 OpenMiningRound can be archived without totals. only the summarizinground contract would block on totals

Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:338]: Discussed with Tim, that due to delay in ingestion in `ScanRewardsReferenceStore` we may not deterministically get the `Some 0` value, and it would block the `SummarizingMiningRoundTrigger`, and `advanceRoundsToNextRoundOpening`. So its a reasonable hack to make the reward computation pipeline work during bootstrap.

Comment (dfordivam) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:338]: it would be good add a comment to explain here why the "Bootstrapped to 0 on a fresh firstSV to make round 0 complete" was needed

Comment (adetokunbo) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala:338]: Done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala [hunk 4]
```diff
@@ -409,8 +423,20 @@ abstract class TrafficBasedRewardsTimeBasedIntegrationTestBase
         }.toMap
       }
 
-    // Rounds 0..4 have no activity records (totals are zero);
-    // rounds 5..10 have activity. Round 11 has not closed yet.
+    clue("Rounds 0..4 have zero activity from bootstrap (no featured apps yet)") {
```
Comment (dfordivam) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala]: Lets move this assertion to the apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/UnhideAndExpireRewardCouponV2TimeBasedIntegrationTest.scala test, as this test pauses verdict ingestion and cannot test whether we can calculate the totals in the abscence of the featured-app activity

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/UnhideAndExpireRewardCouponV2TimeBasedIntegrationTest.scala [hunk 5]
```diff
@@ -162,15 +162,10 @@ class UnhideAndExpireRewardCouponV2TimeBasedIntegrationTest
           c.payload.provider == bobParty.toProtoPrimitive && c.payload.beneficiary.isEmpty
         )
 
-    // TODO: (#5624) Fix the bootstrap of network such that processing of rounds
-    // before the first activity by a featured-app is ingested, and use advanceRoundsToNextRoundOpening
-    for (round <- 1 to 3) {
-      advanceTimeAndWaitForRoundOpening
+    for (round <- 1 to 4) {
+      advanceRoundsToNextRoundOpening
       assertOldestOpenRound(round.toLong)
```
Comment (dfordivam) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/UnhideAndExpireRewardCouponV2TimeBasedIntegrationTest.scala:167]: Lets add the `totalAppActivityWeight` assertion here, before we perform the `doTransfer`

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala (hunk 4 above)
Comment (dfordivam) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala]: Note that the SVs will eventually be able to calculate totals once  featured-app activity starts

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/UnhideAndExpireRewardCouponV2TimeBasedIntegrationTest.scala (hunk 5 above)
Comment (adetokunbo) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/UnhideAndExpireRewardCouponV2TimeBasedIntegrationTest.scala:167]: Done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala (hunk 4 above)
Comment (adetokunbo) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala]: Done

File: apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala [hunk 6]
```diff
@@ -124,24 +124,66 @@ abstract class TrafficBasedRewardsTimeBasedIntegrationTestBase
...
+          Seq(sv2ScanBackend, sv3ScanBackend, sv4ScanBackend).foreach { scan =>
+            scan.getRewardAccountingEarliestAvailableRound() should not be None
+          }
+        }
+      }
+
+      clue("Bootstrap rounds have zero activity (no featured apps yet)") {
+        (0L to 2L).foreach { round =>
+          eventually() {
+            inside(sv1ScanBackend.getRewardAccountingActivityTotals(round)) {
```
Comment (dfordivam) [inline apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala]: Lets confirm round 1 onwards all svs can provide totals

Comment (dfordivam) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala]: And we should ensure that after we do the second `advanceTimeAndWaitForRoundOpening` for closing the round 1, after that immediately we are able to do the totals by all SVs; just to be double sure there is no corner case we are missing.

Comment (adetokunbo) [reply apps/app/src/test/scala/org/lfdecentralizedtrust/splice/integration/tests/TrafficBasedRewardsTimeBasedIntegrationTest.scala]: Fixed

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala [hunk 7]
```diff
@@ -284,49 +280,68 @@ class DbAppActivityRecordStore(
...
       case _ => ()
     }
   }
 
-  /** Insert multiple app activity records in a single transaction.
+  /** Insert activity records only, without meta row management.
+    * Tests manage meta rows separately via `insertActivityRecordMeta`.
     */
   @VisibleForTesting
   def insertAppActivityRecords(
```
Comment (meiersi-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala]:
```suggestion
  def insertAppActivityRecordsForTesting(
```

Make the name such that it is unlikely that it is used in prod where AFAIU it would be wrong to use.

Comment (meiersi-da) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala]: Also apply to other functions that should only be used for testing and could be mistaken for the real thing.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/ScanVerdictIngestionService.scala [hunk 8]
```diff
@@ -264,7 +265,7 @@ class ScanVerdictIngestionService(
 
         _ <- ensureVerdictsHaveTrafficSummaries(verdicts, summaryByTime)
         _ <- store.insertVerdictsWithAppActivityRecords(
-          items,
+          NonEmptyList.fromListUnsafe(items.toList),
```
Comment (meiersi-da) [inline apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/ScanVerdictIngestionService.scala]: nit: the nicer option would be to establish the non-emptiness on `verdicts` using a pattern match and thereby make the `map` produce a non-empty list of `items`.

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/scan/store/DbScanAppRewardsStoreTest.scala [hunk 9]
```diff
@@ -456,7 +456,7 @@ class DbScanAppRewardsStoreTest
 
     "aggregateActivityTotals — succeeds for round 0 when isFirstSv" in {
       for {
-        (store, historyId) <- newStore(isFirstSv = true)
+        (store, historyId) <- newStore()
```
Comment (meiersi-da) [inline apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/scan/store/DbScanAppRewardsStoreTest.scala]: comment and store initialization don't seem to match. What are you testing / did you want to test?

**Review by meiersi-da — APPROVED** [review]
Nice. Thanks.

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala (hunk 7 above)
Comment (adetokunbo) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/db/DbAppActivityRecordStore.scala]: Fixed

File: apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/ScanVerdictIngestionService.scala (hunk 8 above)
Comment (adetokunbo) [reply apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/automation/ScanVerdictIngestionService.scala]: Fixed

File: apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/scan/store/DbScanAppRewardsStoreTest.scala (hunk 9 above)
Comment (adetokunbo) [reply apps/scan/src/test/scala/org/lfdecentralizedtrust/splice/scan/store/DbScanAppRewardsStoreTest.scala]: Updated the test description, isFirstSv was not under test here
